src/odin/ingestion/spare/spare_job.py

```diff
@@ -0,0 +1,400 @@
+    # we need the id, but the other columns don't matter.
+    # this sets other cols to the last value before it was deleted.
+    soft_deletes = compacted_prev.filter(pl.col("id").is_in(deleted_ids)).with_columns(
+        pl.lit(datetime_ms, pl.Datetime("ms")).alias("ROW_TIMESTAMP"),
+        pl.lit(True, pl.Boolean()).alias("ROW_DELETED"),
+    )
+    new_rows = new_df.filter(pl.col("id").is_in(new_and_changed_ids))
+
+    if soft_deletes.is_empty() and new_rows.is_empty():
+        return prev_df
+    else:
+        return pl.concat([prev_df, soft_deletes, new_rows], how="vertical")
+
+
+def run(
+    config: Config,
+    api_endpoint: ApiEndpoint,
+) -> None:
+    """
+    Run the job once.
+
+    In a function separate from the job class so it can be tested.
+    """
+    validate_config(config)
+    source = config["source"]
+    target = config["target"]
+    columns = api_endpoint["columns"]
+    parquet_path = os.path.join(target, f"{api_endpoint['name']}.parquet")
+
+    # get new input files
+    paths = spare_s3.list_objects(source)
+    paths = filter(lambda path: api_endpoint["api_path"] in path, paths)
+    # process new files in chronological order (which is also alphabetical order)
+    paths = sorted(paths)
+    ProcessLog(
+        "SpareJob get_inputs",
+        api_name=api_endpoint["name"],
+        source=source,
+        api_path=api_endpoint["api_path"],
+        num_paths=len(paths),
+    )
+    if paths == []:
+        # no new files to process
+        return
+
+    df = load_parquet(parquet_path, columns)
+    rows_before = len(df)
+    for path in paths:
+        with spare_s3.open_file(os.path.join(source, path)) as f:
+            input_df = load_json(f, columns)
+        # add metadata
+        datetime_ms = datetime_from_delta_path_ms(path)
+        rows_input = len(input_df)
+        # add new metadata columns as constants to the left
+        input_df = input_df.select(
+            pl.lit(datetime_ms, pl.Datetime("ms")).alias("ROW_TIMESTAMP"),
+            pl.lit(False, pl.Boolean()).alias("ROW_DELETED"),
+            pl.all(),
+        )
+        rows_before_merge = len(df)
+        df = merge_new_data(df, input_df, columns, datetime_ms)
+        rows_after_merge = len(df)
+        # log: input name, num rows, num rows in prev, num rows after. also api_name
+        ProcessLog(
+            "SpareJob process_input",
+            api_name=api_endpoint["name"],
+            obj_path=path,
+            rows_input=rows_input,
+            rows_before_merge=rows_before_merge,
+            rows_after_merge=rows_after_merge,
+        )
+
+    rows_after = len(df)
+    should_write_parquet = rows_after > rows_before
+    ProcessLog(
+        "SpareJob write_parquet",
+        api_name=api_endpoint["name"],
+        write_parquet=should_write_parquet,
+        rows_before=rows_before,
+        rows_after=rows_after,
+        parquet_path=parquet_path,
+    )
+    if should_write_parquet:
```

Comment:
It looks like if you end up processing files where no records change, and only `ROW_TIMESTMAP` is updated, they won't be written back out to the parquet file.

Reply:
Yeah, that's intentional. Most of the time, most data in this dataset isn't changing. Skipping the write means the file won't grow larger with rows that don't contain new information. The tradeoff is that there's no timestamp in the file indicating how long it's been since it polled for updates, but I think that's fine.